### PR TITLE
[WIP] rust-docs for External

### PR DESCRIPTION
**Pull request by fckt**

### CODE REVIEW

File: core/primitives/src/transaction.rs [hunk 1]
```diff
@@ -104,6 +104,7 @@ pub struct TransferAction {
 #[derive(BorshSerialize, BorshDeserialize, PartialEq, Eq, Clone, Debug)]
 pub struct StakeAction {
     pub stake: Balance,
+    /// A public key of a validator
```
Comment (bowenwang1996) [inline core/primitives/src/transaction.rs]: This comment seems unnecessary/confusing. Maybe say that it is a staking key/validator key

File: core/store/src/trie/update.rs [hunk 2]
```diff
@@ -394,5 +394,8 @@ mod tests {
 
         let values: Vec<Vec<u8>> = trie_update.range(b"do", b"", b"xyz").unwrap().collect();
         assert_eq!(values, vec![b"dog".to_vec(), b"dog2".to_vec(), b"dog3".to_vec()]);
+
+        let values: Vec<Vec<u8>> = trie_update.range(b"do", b"bb", b"12").unwrap().collect();
+        // assert_eq!(values, vec![b"dog".to_vec(), b"dog2".to_vec(), b"dog3".to_vec()]);
```
Comment (bowenwang1996) [inline core/store/src/trie/update.rs]: why is this commented out?

Comment (fckt) [reply core/store/src/trie/update.rs]: fixed

File: runtime/near-vm-logic/src/dependencies.rs [hunk 3]
```diff
@@ -32,40 +32,285 @@ pub trait MemoryLike {
...
 
+    /// Creates a receipt which will be executed after `receipt_indices`
+    ///
+    /// # Arguments
+    ///
+    /// * `receipt_indices` - a list of receipt indices the new receipt is depend on
+    ///
+    /// # Example
+    /// ```
+    /// use near_vm_logic::mocks::mock_external::MockedExternal;
```
Comment (nearmax) [inline runtime/near-vm-logic/src/dependencies.rs]: I would hide the repeating parts like these from the docstring here and everywhere else.
```rust
use near_vm_logic::External;
let mut external = MockedExternal::new();
```
 https://doc.rust-lang.org/rustdoc/documentation-tests.html#hiding-portions-of-the-example

**Comment by nearmax** [comment]
@ailisp Check this out, @fckt has added the comments (and tests) but the coverage has decreased blocking the merge.

**Comment by ailisp** [comment]
Looks like all `///` comment is not correctly handled by kcov. Looking at its help there's
```
 --exclude-line=pat[,...] Consider lines that match the patterns to be non-
                         code lines.
```
Let me try to exclude these lines

**Comment by ailisp** [comment]
Looks it's not kcov's fault. For example primitives/challenge.rs:
image
kcov report a 90%+ coverage. `///` lines are ignored. According to codecov: https://docs.codecov.io/docs/merging-reports it suppose to merge coverage from different test so that: If one line is covered in one of all test, it should be marked as covered. However this is false, from the codecov report from this PR:
https://codecov.io/gh/nearprotocol/nearcore/pull/1780/src/core/primitives/src/challenge.rs


Also I'm strongly suspect the quality of codecov report on a rust project. Looking the history coverage of this file, it looks pretty random:
https://codecov.io/gh/nearprotocol/nearcore/branch/master/history/core/primitives/src/challenge.rs?before=cff11dbd3356487871dccb9f7948ee634c3d7e0e

**Comment by ailisp** [comment]
kcov comes with a `--merge` feature, which i tested gives 91% after merge one 91% and one 0% challenge.rs report. I'll merge and upload only merged report to codecov.

**Pull request merged**